compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+            }
+            "Result" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg_ok = self.print(args.get(0).expect("print_prelude_type result ok"));
+                let arg_err = self.print(args.get(1).expect("print_prelude_type result err"));
+                docvec!["$gleam.Result<", arg_ok, ", ", arg_err, ">"]
+            }
+            // Getting here sholud mean we either forgot a built-in type or there is a
+            // compiler error
+            name => panic!("{} is not a built-in type.", name),
+        }
+    }
+
+    /// Prints a "named" programmer-defined Gleam type into the Typescript
+    /// equivalent.
+    ///
+    fn print_type_app(
+        &self,
+        name: &str,
+        args: &[Arc<Type>],
+        module: &[String],
+    ) -> Document<'static> {
+        let name = format!("{}$", ts_safe_type_name(name.to_camel_case()));
+        let name = match module == self.current_module {
+            true => Document::String(name),
+            false => {
+                // If type comes from a seperate module, use that module's nam
+                // as a Typescript namespace prefix
+                let module_name = module
+                    .last()
+                    .expect("Typescript generator could not identify imported module name.");
+                let module_name = format!("${}", module_name);
+                docvec![Document::String(module_name), ".", Document::String(name),]
+            }
+        };
+        if args.is_empty() {
+            return name;
+        }
+        // If the App type takes arguments, pass them in as Typescript generics
+        let args = concat(Itertools::intersperse(
+            args.iter().map(|a| self.print(a)),
+            ", ".to_doc(),
+        ));
+        docvec![name, "<", args, ">"]
+    }
+
+    /// Prints the Typescript type for an anonymous function (aka lambda)
+    ///
+    fn print_fn(&self, args: &[Arc<Type>], retrn: &Type) -> Document<'static> {
+        let args = concat(Itertools::intersperse(
+            args.iter()
+                .enumerate()
+                .map(|(idx, a)| docvec!["x", idx, ": ", self.print(a)]),
+            ", ".to_doc(),
+        ));
+        let retrn = self.print(retrn);
+        docvec!["(", args, ") => ", retrn]
+    }
+
+    /// Allows an outside module to mark the Gleam prelude as "used"
+    ///
+    pub fn set_prelude_used(&self) {
+        self.tracker.borrow_mut().prelude_used = true
+    }
+
+    /// Returns if the Gleam prelude has been used at all during the process
+    /// of printing the Typescript types
+    ///
+    pub fn prelude_used(&self) -> bool {
+        self.tracker.borrow().prelude_used
+    }
+}
+
+/// Prints a Gleam tuple in the Typescript equivalent syntax
+///
+fn tuple<'a>(elems: impl IntoIterator<Item = Document<'a>>) -> Document<'a> {
+    concat(Itertools::intersperse(elems.into_iter(), break_(",", ", ")))
+        .nest_current()
```

Comment:
Can we indent by 1 rather than using `nest_current`? I think that would suit the TypeScript style more.

It would also be good to have a test for how a really long tuple renders, to ensure it is formatted correctly.